### ISSUE TITLE
Reduce memory footprint

### ISSUE THREAD

**Issue by rhc54** [pmix/master#129]
Fujitsu has kindly shared their memory footprint measurements at scale, and the PMIx footprint is growing at an unacceptable rate. The primary reasons are two-fold:
1. the use of fixed-length keys and namespace values is consuming a lot of empty space.
2. presence of a pmix_proc_t struct, which in turn contains a fixed-length namespace array, in the pmix_value_t data union, thereby causing that oft-used struct to be quite large

Concern was raised over the use of these in the pmix_info_t structure, but that is a red-herring. PMIx only stores pmix_kval_t objects - the pmix_info_t structure is only used for interfacing to the outside. Thus, we can resolve the problem by:
1. changing the pmix_kval_t to use flexible sized key strings. To avoid multiple tiny malloc's, we will move the key entry to the end of the struct and pre-compute size using the known string length.
2. replacing the pmix_proc_t in the pmix_value_t struct with a pointer to that struct
3. replacing fixed namespace arrays with variable-length storage in all internal objects

These steps will resolve the problem without impacting any of the RM's or applications that have implemented their integration. This is a good thing as that would have been highly disruptive.

**Comment by kawashima-fj** [comment]
The presented material:
https://github.com/open-mpi/ompi/wiki/Meeting-2016-08-attachments/Fujitsu-OMPI-Dev-Meeting-2016-Aug.pdf
(also linked from [Open MPI Developer's meeting Wiki page](https://github.com/open-mpi/ompi/wiki/Meeting-2016-08))

**Comment by jjhursey** [comment]
We found a memory leak in PMIx_Get in the server side code a couple weeks ago. We have a fix that we need to push upstream - @dsolt and @nysal did that work. It might not be a huge factor here (depending on the number of get calls), but would be good to push back.

**Comment by rhc54** [comment]
Well, I went thru the code base and found that we already use pmix\_kval\_t everywhere internally, converting all pmix\_info\_t values to that struct. Thus, the first part of my list of fixes was already done, and nothing to be gained there.

So I did the second part in the referenced PR, squashing the pmix\_value\_t struct down to 8 bytes, and committed that to the OMPI master so folks can see how that impacts the memory footprint. I also enabled the shared memory datastore, which should also help.

@hjelmn said he could test this next week, so let's see what this now looks like.

**Comment by kawashima-fj** [comment]
@rhc54 Thanks! I'll evaluate the memory footprint again using the latest master. The result will be posted here in late this week hopefully.

**Comment by artpol84** [comment]
@kawashima-fj can you provide mca params used when add_proc was disabled?
if you only specified
```
--mca mpi_add_procs_cutoff 0
```
without setting
```
--mca pmix_base_async_modex 1
```
you'll still end up doing the full modex and you will see all keys exchanged and stored inside **every** pmix server.

**Comment by kawashima-fj** [comment]
@artpol84 Sorry, I missed your comment. I used the following MCA parameters.

For **dynamic add_procs enabled** case (a):
```
--mca btl self,dummy
--mca pmix_base_collect_data 0
```
(because the default value of `mpi_add_procs_cutoff` is 0)

For **dynamic add_procs disabled** case (b):
```
--mca btl self,dummy
--mca pmix_base_collect_data 0
--mca mpi_add_procs_cutoff 99999
```

Today I added `--mca pmix_base_async_modex 1` (c):
```
--mca btl self,dummy
--mca pmix_base_collect_data 0
--mca pmix_base_async_modex 1
```
and I measured the memory footprint of (a) and (c).

Memory footprint of (c) (84.6MiB/node for nnodes=48 and ppn=16) was larger than (a) (80.4MiB/node for nnodes=48 and ppn=16). Is this an expected result? It  seems odd. I used Open MPI master 277c319 (2016-08-26 09:42:45 +0900) and embedded PMIx 2.0.

**Comment by kawashima-fj** [comment]
I measured memory footprint using recent OMPI/PMIx.

Evaluation environment/condition is same as one written in [my OMPI meeting material](https://github.com/open-mpi/ompi/wiki/Meeting-2016-08-attachments/Fujitsu-OMPI-Dev-Meeting-2016-Aug.pdf). `pmix_base_async_modex` suggested by @artpol84 is not used (I missed the comment and looked a little while ago, sorry).

(a) Open MPI 2.0.0 + PMIx 1.1.2 embedded in OMPI (same as P.6 in [my OMPI meeting material](https://github.com/open-mpi/ompi/wiki/Meeting-2016-08-attachments/Fujitsu-OMPI-Dev-Meeting-2016-Aug.pdf))

![graph 200](https://cloud.githubusercontent.com/assets/12043176/18051964/b8b9a8d4-6e32-11e6-9ce5-09406700e83f.png)

(b) Open MPI master bbc6d4b (22 July) + PMIx master b7120bd (18 July) (same as P.7 in [my OMPI meeting material](https://github.com/open-mpi/ompi/wiki/Meeting-2016-08-attachments/Fujitsu-OMPI-Dev-Meeting-2016-Aug.pdf))

![graph july](https://cloud.githubusercontent.com/assets/12043176/18051969/bdcdc030-6e32-11e6-8580-e3fc6c245fdb.png)

(c) Open MPI master 277c319 (26 Aug.) + PMIx 2.0a embedded in OMPI

![graph aug2](https://cloud.githubusercontent.com/assets/12043176/18051972/c2e91132-6e32-11e6-9294-c9b3be0b297e.png)

(d) Open MPI master 277c319 (26 Aug.) + PMIx master a23771e (26 Aug.)

![graph augm](https://cloud.githubusercontent.com/assets/12043176/18051975/c631a76e-6e32-11e6-8038-62893b38f65d.png)

Memory footprint of PMIx client part (area between blue and red lines) is reduced but that of PMIx server part (red line) is increced. Is this the effect of shared memory dstore?

The measured memory is the amount of physical memory that the job obtained from kernel.

**Comment by rhc54** [comment]
I'm a little confused here. I gather the red line is the size of the PMIx server, but that should change under the 3 use-cases you show for the client. Are you seeing the server being the same size in all cases? Is there some meaning to the "sleep 5s" label for that line?

Either way, it appears that this confirms what I had feared - the changes I made simply shifted memory from the client to the server, but had no significant benefit. I think we need to scrub what is being "put" by the clients, and maybe come up with a more efficient representation for data instead of just storing every individual data point.

**Comment by kawashima-fj** [comment]
@rhc54 Sorry, my explanation was inaccurate.

I executed four jobs (in PMIx terms, 'four sessions') and each line in the graph shows memory consumption of each job.

The red line is `mpiexec -n N sleep 5s`. So the line shows memory consunption of one `mpiexec` process (**without** any PMIx client API interaction) and PPN `sleep` processes. I assume here that memory consumption of `sleep` process can be ignored in a large-scale run.

The blue line is `mpiexec -n N ./pmix-init-commit-fence-finalize`. So the line shows memory consunption of one `mpiexec` process (**with** PMIx client API interaction) and PPN PMIx client processes.

The delta of the blue line and the red line consists of memory used for PMIx client API interaction in the `mpiexec` process and memory used in the PMIx client processes.

**Comment by rhc54** [comment]
@kawashima-fj Thank you for the explanation - that helped a lot! It appears we have actually gotten worse with the shared memory as our total memory consumption increased when compared to your earlier measurements. This will take some thought as to why it might have happened as there are other changes that also occurred.

**Comment by artpol84** [comment]
What about the blue line:
PMIx_{Init,Commit,Fence,Finalize} - is this a raw PMIx program? If so - what Fence was used - with or without data collection?
Can you point us to exact programs that were used?

**Comment by kawashima-fj** [comment]
@artpol84 It's a raw PMIx program without PMIx_Put.

```c
#include <stdio.h>
#include <pmix.h>

#define CALL(call)                                     \
    do {                                               \
        pmix_status_t status = call;                   \
        if (status != PMIX_SUCCESS) {                  \
            fprintf(stderr, "%s = %d", #call, status); \
        }                                              \
    } while (0)

int main(int argc, char *argv[])
{
    pmix_proc_t proc;
    CALL(PMIx_Init(&proc, NULL, 0));
    CALL(PMIx_Commit());
    CALL(PMIx_Fence(NULL, 0, NULL, 0));
    CALL(PMIx_Finalize(NULL, 0));
    return 0;
}
```

**Comment by rhc54** [comment]
I'm wondering if we are measuring what we think we are - let me explain. The PMIx server will use the same amount of memory for registering namespaces and clients regardless of whether or not the client calls PMIx_Init. Thus, minus any modex data, the size of the server must be exactly the same under any of these test cases.

In the latest versions, we have the PMIx shared memory enabled by default. Thus, the clients no longer receive a copy of the namespace data - they simply read it from the server's shared memory. So the size of the client cannot be a function of the size of the launched job for the PMIx-only program. There are no per-peer structures being created (outside of the one to the local server), and thus no memory storage is being consumed.

Therefore, I believe we are getting confused by these graphs. What I think we should do is use OMPI's ability to report the actual memory consumption of each process, both the daemons and the clients instead of looking at the total memory consumed on the node. I'll check to ensure it still works (haven't used it in a while), and then send along some instructions.

**Comment by rhc54** [comment]
Okay, I have a few things to report. First, and most important, I believe we are either not measuring memory usage correctly, or we need to understand how to properly interpret the data the Fujitsu system is reporting. I did some digging and found, as you might expect, a _lot_ of discussion out there about how to measure actual memory usage. It is a complex subject due to shared libraries and virtual memory.

The most consistent determination I could locate was to use the PSS (Proportional Set Size). This adds together the unique set size (USS) with a proportion of the shared memory divided across the number of processes sharing that memory on the node. Thus, you get an idea of the physical memory footprint induced by that process.

The best method I found for getting that info was to use the [smem](http://manpages.ubuntu.com/manpages/precise/en/man8/smem.8.html) command. It does a very nice job of outputting the info for each process being executed on the node. Nothing magic about the method employed - it simply parses the /proc/pid/smaps file and combines the relevant data, then pretty-prints it. I'll look at doing something similar in orte-top.

Meantime, what I found would indicate that we really may not have much of a problem. Here is a quick table of results - note that I only have a 2-node system right now, so this obviously needs to be replicated at scale:

```table
#procs/node |    mpirun    |    orted      | mpi_spin (HNP) |  mpi_spin(orted)
       2    |     12.2M    |    11.5M      |        4.6M    |        4.7M
       4    |     12.0M    |    11.3M      |        4.3M    |        4.4M 
       8    |     11.9M    |    11.1M      |        4.0M    |        4.0M
      16    |     12.1M    |    11.3M      |        3.9M    |        3.9M
      24    |     12.5M    |    11.7M      |        3.8M    |        3.9M
```
These measurements were taken using the current OMPI master, which has an up-to-date version of the PMIx master in it. I used a simple ```mpirun -npernode N ./mpi_spin``` command, so there is a full data exchange in the modex, with each proc contributing its local modex data. This is a TCP-only system, so that includes the shared memory rendezvous point (that only is shared to other procs on the same node - it isn't circulated) and the TCP/BTL socket.

There is some variation in the per-process PSS output as the algorithm attempts to divide the shared libraries evenly, but eventually has to assign any leftover to someone. Still, this provides at least a uniform method of measurement that we should all be able to replicate as it doesn't depend upon any specific environment (other than a flavor of POSIX).

As I would have expected, the use of the PMIx shared memory storage in the server causes the MPI procs to remain at a stable size as the size of the job grows. This is because the mpi_spin application calls MPI_Init, but never attempts to communicate - it just goes into a tight loop doing nothing useful. Thus, no connection information is ever retrieved, and no MPI proc structs are created.

The daemon grows a little with the size of the job because it is carrying all the job-level info, and that is proportional to the number of nodes and procs. Note, however, that it isn't growing rapidly, so the footprint/proc looks like it could be small. We need data at much larger scales to confirm this notion.

I will add a memory profiling option to ORTE that will allow us to get the PSS for the daemon and child processes so this measurement will be easy to perform whenever we want to do so.

**Comment by kawashima-fj** [comment]
Interesting. I used the Fujitsu RM feature to know memory footprint but it calculates shared memory in a different way (I don't know the detail). I can measure memory footprint up to ~256 nodes with the memory profiling option once it is implemented.

**Comment by artpol84** [comment]
Maybe it accounts shared memory piece at each proc that mapped it multiple times?

I agree with @rhc54 that we need an uniform way to measure that so we all can debug/reproduce this.

**Comment by artpol84** [comment]
I mean same shmem piece accounted multiple times because multiple procs map it.

**Comment by rhc54** [comment]
Working its way thru Jenkins now:

https://github.com/open-mpi/ompi/pull/2035

**Comment by rhc54** [comment]
This has been committed to OMPI master now. It builds by default, and you activate it by simply setting OMPI_MEMPROFILE=N in your environment. The N is the number of seconds you want to wait before ORTE samples the memory usage. Once ORTE does the sample, it will output the results and then terminate the job.

You can do this with any application. The only real requirement is that you need to have the app "spin" at the point where you want to measure, and ensure that the OMPI_MEMPROFILE value is set to a long enough time to ensure your procs have reached that point. So my mpi_spin test (located in orte/test/mpi) looks like this:

```c
/* -*- C -*-
 *
 * $HEADER$
 *
 * A program that just spins - provides mechanism for testing user-driven
 * abnormal program termination
 */

#include <stdio.h>
#include "mpi.h"

int main(int argc, char* argv[])
{

    int i;
    double pi;

    MPI_Init(&argc, &argv);

    i = 0;
    while (1) {
        i++;
        pi = i / 3.14159256;
        if (i > 100) i = 0;
    }

    MPI_Finalize();

    return 0;
}
```

This allows me to measure the memory footprint after the modex has completed. The output consists of:

* footprint of mpirun itself, plus the average of all its local clients
* footprint of the daemon on the next remote node, plus the average of all its local clients

I didn't have it report for every node as that seemed overly redundant, but we can set it up for that if folks want.

**Comment by kawashima-fj** [comment]
I evaluated the memory footprint using `OMPI_MEMPROFILE` feature. The result is listed below.

### Result

```
ppn: 16

-----------------------------------------------------------------------------

prog: sleep command

#node |  mpirun |  orted  | prog(HNP) | prog(orted) | node(HNP) | node(orted)
  32  |   9.12M |   8.56M |   0.40M   |    0.42M    |   15.52M  |   15.28M
  64  |  14.28M |  13.34M |   0.40M   |    0.41M    |   20.68M  |   19.90M
  96  |  19.27M |  18.29M |   0.40M   |    0.41M    |   25.67M  |   24.85M
 128  |  24.33M |  23.13M |   0.40M   |    0.41M    |   30.73M  |   29.69M
 160  |  29.24M |  28.03M |   0.40M   |    0.41M    |   35.64M  |   34.59M
 192  |  34.37M |  32.72M |   0.40M   |    0.41M    |   40.77M  |   39.28M

-----------------------------------------------------------------------------

prog: PMIx program (PMIx_{Init,Commit,Fence},spin_loop,PMIx_Finalize)

#node |  mpirun |  orted  | prog(HNP) | prog(orted) | node(HNP) | node(orted)
  32  |   9.51M |   8.67M |   1.43M   |    1.43M    |   32.39M  |   31.55M
  64  |  14.74M |  13.93M |   2.22M   |    2.23M    |   50.26M  |   49.61M
  96  |  20.06M |  19.35M |   3.07M   |    3.08M    |   69.18M  |   68.63M
 128  |  26.83M |  24.69M |   4.01M   |    4.03M    |   90.99M  |   89.17M
 160  |  31.54M |  30.05M |   4.83M   |    4.85M    |  108.82M  |  107.65M
 192  |  37.07M |  35.24M |   5.79M   |    5.80M    |  129.71M  |  128.04M

-----------------------------------------------------------------------------

prog: MPI program (MPI_Init,spin_loop,MPI_Finalize)

#node |  mpirun |  orted  | prog(HNP) | prog(orted) | node(HNP) | node(orted)
  32  |   8.28M |   7.32M |   3.19M   |    3.27M    |   59.32M  |   59.64M
  64  |  13.66M |  12.60M |   4.36M   |    4.40M    |   83.42M  |   83.00M
  96  |  19.73M |  18.75M |   8.55M   |    5.59M    |  156.53M  |  108.19M
 128  |  24.69M |  25.31M |   6.79M   |    6.83M    |  133.33M  |  134.59M
 160  |  29.96M |  28.75M |   8.03M   |    8.08M    |  158.44M  |  158.03M
 192  |  37.01M |  35.38M |   9.23M   |    9.32M    |  184.69M  |  184.50M
```

The last two columns show the memory footprint per node. `node(HNP)` is `mpirun + prog(HNP) * 16` and `node(orted)` is `orted + prog(orted) * 16` (16 is ppn).

### Condition

ppn: 16

Open MPI: master fb51d65 (Sept. 5)
PMIx: master 1f7cccb (Sept. 5) + my patch #140

mpiexec options:

```
-n $nprocs
--hostfile $hostfile
--mca btl self,dummy                                           
--mca pmix pmix3x
--mca pmix_base_collect_data 0
--mca pmix_base_async_modex 1
--mca mpi_add_procs_cutoff 99999
-x OMPI_MEMPROFILE=30
```

The PMIx program spins forever after calling `PMIx_Fence`. The MPI program spins forever after calling `MPI_Init`. The value `OMPI_MEMPROFILE=30` is enough to complete the initialzation process. I confirmed it by inserting `printf` before the spin loop.

It can be confirmed that the shared memory data store is used because a bus error occurs if my patch #140 is not applied.

### Discussion

This result shows the memory footprint scales with the number of nodes (processes). The values are not same as the ones in my graph shown above but similar. (Note that the number of node is up to 256 in my graph but it is up to 192 this time.)

I read the latest PMIx code and noticed that data passed to `PMIx_server_register_nspace` are not on the shared memory datastore.

I put `printf` at the bottom of `_register_nspace` function to look the amount of data.

```diff
--- a/src/server/pmix_server.c
+++ b/src/server/pmix_server.c
@@ -720,6 +720,9 @@ static void _register_nspace(int sd, short args, void *cbdata)
     }
     /* do not destruct the kv object - no memory leak will result */
 
+    printf("server job_info: %zu\n", nptr->server->job_info.bytes_used);
+    fflush(stdout);
+
 #if defined(PMIX_ENABLE_DSTORE) && (PMIX_ENABLE_DSTORE == 1)
     if (0 > pmix_dstore_nspace_add(cd->proc.nspace)) {
         PMIX_ERROR_LOG(rc);
```

The result for ppn=16 is:

```
 #node | job_info size
   64  |   489,553
  128  |   997,447
  192  | 1,505,443
```

These job data are stored in the PMIx server. They are passed to all PMIx clients via usock when PMIx clients connect to the PMIx server and they are converted to lists of `pmix_kval_t` on the clients. So all server/clients processes have these data in their independent memory. For example of #node=192, at least 1,505,443 * 17 = 24.4MiB memory is consumed per node. While converting `pmix_buffer_t` to lists of `pmix_kval_t`, both data exist in memory. So double memory are allocated and will be counted in `/proc/pid/smaps` if libc `free` does not return the memory to the kernel immediately.

I'm not sure the `job_info` is the main part of the memory footprint but it seems one part.

**Comment by artpol84** [comment]
@kawashima-fj 
```
--mca mpi_add_procs_cutoff 99999
```
I guess you want to set cutoff to 0 otherwise you trigger addproc for each process.

**Comment by artpol84** [comment]
And this explains why you see dstor usage.

**Comment by artpol84** [comment]
And you are right about job info - we need to provide it through dstor as well.

**Comment by kawashima-fj** [comment]
Oh, my bad. I wanted to set it to 0. I reused a wrong shell script, sigh.

But I believe it affects only MPI programs. The result of a pure PMIx program is still valid. Is it?

**Comment by artpol84** [comment]
Yes, this is only for MPI progs.

**Comment by artpol84** [comment]
And my believe is that job info is causing problems. Some keys IIRC scale along with number of procs or nodes:
* `PMIX_NODE_LIST/PMIX_ALLOCATED_NODELIST` (if not compressed) will scale as # of nodes. 
Btw @rhc54 I see the duplication of `PMIX_NODE_LIST`:
  * https://github.com/pmix/master/blob/master/include/pmix_common.h#L154
  * https://github.com/pmix/master/blob/master/include/pmix_common.h#L177
* `PMIX_PROC_MAP` - *regex describing procs on each node within this job* - unless it is packed in a smart way will scale as # of processes. 
* `PMIX_MAP_BLOB` i guess this contains the same info as `PMIX_PROC_MAP` but in different representation. If so - it may also scale as # of procs (@rhc54, please comment).

So in worst case we have `nproc * nproc` scaling for some of the keys (because each proc will have the key scaling as `nproc`).
If we will put job info into dstore we will reduce the scaling effect to be `nprocs * nnodes` which might be sufficient for a while.

However in general it would be good to think of the ways to decrease this dependency and let direct modex to solve this. However it's not an easy thing to do because to fetch the data for a particular process with direct modex we at least need to know on what node this process is located.

**Comment by karasevb** [comment]
We make sure the `job_info` is not use shared memory datastore. 
I was debug a simple program which  uses `PMIx_Get` for any keys contained in `job_info`. This allowed us to see that `PMIx_server_register_nspace` pass `job_info` to every process and every process is store its own copy of `job_info` in local hash. 
If will be using dstor for `job_info` it will reduce the amount of memory for this case.

**Comment by jsquyres** [comment]
Today, I ran the `examples/ring_c` through a memory leak checker on the OMPI master (PMIx 3.x).  It turned up a bunch of little leaks.  I've attached a report showing what I think are just the PMIx-specific leaks (I'll file a separate report for the OMPI leaks).

Is this information useful here on this issue, or does the PMIx community want this filed on a separate issue?

[results-pmix.txt](https://github.com/pmix/master/files/509565/results-pmix.txt)
